### DISCUSSION TITLE
docker build error ENOENT: no such file or directory, open '/app/src/router/index.js'

### DISCUSSION THREAD

**Discussion by gjtjx** [quasarframework/quasar#16124] [CLI - Electron mode]
when i run sudo docker build --no-cache -t gokalpgoren/webapp:v1.0.1-arm64v8 . in feniks.web_app
I have got error like this
Sending build context to Docker daemon 1.399MB
Step 1/23 : FROM node:latest as develop-stage
---> b1e43aea8c4f
Step 2/23 : WORKDIR /app
---> Using cache
---> 03f09120c7cb
Step 3/23 : COPY electron-v7.3.3-linux-arm64.zip /app3/node_modules/electron/electron/releases/download/v7.3.3/
---> Using cache
---> 80983d7ad5d8
Step 4/23 : COPY SHASUMS256.txt /app3/node_modules/electron/electron/releases/download/v7.3.3/
---> Using cache
---> 765fb8a57b2f
Step 5/23 : COPY package*.json ./
---> Using cache
---> c4647782216a
Step 6/23 : COPY . .
---> 21932e871128
Step 7/23 : RUN ls /app/src/router -al
---> Running in c9d1d6ab22e8
total 16
drwxrwxr-x 2 root root 4096 Jul 26 15:31 .
drwxrwxr-x 12 root root 4096 Jul 28 03:35 ..
-rw-rw-r-- 1 root root 1029 Jul 26 15:31 index.ts
-rw-rw-r-- 1 root root 1040 Jul 26 15:31 routes.ts
Removing intermediate container c9d1d6ab22e8
---> 25d7f25f969c
Step 8/23 : FROM develop-stage as build-stage
---> 25d7f25f969c
Step 9/23 : RUN npm config set registry http://registry.npmmirror.com/
---> Running in cb18d9e1596d
Removing intermediate container cb18d9e1596d
---> 93a6580ea33a
Step 10/23 : RUN export ELECTRON_MIRROR=$ELECTRON_MIRROR:http://npm.taobao.org/mirrors/electron/
---> Running in 36b01ba67dbd
Removing intermediate container 36b01ba67dbd
---> 3d39fdf74365
Step 11/23 : RUN export ELECTRON_CONFIG_CACHE=$ELECTRON_CONFIG_CACHE:/app3/node_modules/electron/electron/releases/download/v7.3.3/
---> Running in c3c8c6fa980f
Removing intermediate container c3c8c6fa980f
---> 7f75e382b514
Step 12/23 : RUN npm --registry=http://registry.npmmirror.com/ install
---> Running in b7090f4aff42
npm WARN deprecated stable@0.1.8: Modern JS already guarantees Array#sort() is a stable sort, so this library is deprecated. See the compatibility table on MDN: https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Array/sort#browser_compatibility
npm WARN deprecated debug@4.1.1: Debug versions >=3.2.0 <3.2.7 || >=4 <4.3.1 have a low-severity ReDos regression when used in a Node.js environment. It is recommended you upgrade to 3.2.7 or 4.3.1. (https://github.com/debug-js/debug/issues/797)
npm WARN deprecated debug@4.1.1: Debug versions >=3.2.0 <3.2.7 || >=4 <4.3.1 have a low-severity ReDos regression when used in a Node.js environment. It is recommended you upgrade to 3.2.7 or 4.3.1. (https://github.com/debug-js/debug/issues/797)
npm WARN deprecated debug@4.1.1: Debug versions >=3.2.0 <3.2.7 || >=4 <4.3.1 have a low-severity ReDos regression when used in a Node.js environment. It is recommended you upgrade to 3.2.7 or 4.3.1. (https://github.com/debug-js/debug/issues/797)

added 1149 packages in 32m
npm notice
npm notice New patch version of npm available! 9.8.0 -> 9.8.1
npm notice Changelog: https://github.com/npm/cli/releases/tag/v9.8.1
npm notice Run npm install -g npm@9.8.1 to update!
npm notice
Removing intermediate container b7090f4aff42
---> e2e7fefe4344
Step 13/23 : RUN npm config get registry
---> Running in 02f581371ea4
http://registry.npmmirror.com/
Removing intermediate container 02f581371ea4
---> 2cb407feb087
Step 14/23 : RUN npm install -g quasar-cli
---> Running in db55646cd7b6
npm WARN deprecated stable@0.1.8: Modern JS already guarantees Array#sort() is a stable sort, so this library is deprecated. See the compatibility table on MDN: https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Array/sort#browser_compatibility
npm WARN deprecated topo@2.0.2: This version has been deprecated in accordance with the hapi support policy (hapi.im/support). Please upgrade to the latest version to get the best features, bug fixes, and security patches. If you are unable to upgrade at this time, paid support is available for older versions (hapi.im/commercial).
npm WARN deprecated hoek@4.2.1: This version has been deprecated in accordance with the hapi support policy (hapi.im/support). Please upgrade to the latest version to get the best features, bug fixes, and security patches. If you are unable to upgrade at this time, paid support is available for older versions (hapi.im/commercial).
npm WARN deprecated chokidar@2.1.8: Chokidar 2 does not receive security updates since 2019. Upgrade to chokidar 3 with 15x fewer dependencies
npm WARN deprecated uuid@3.4.0: Please upgrade to version 7 or higher. Older versions may use Math.random() in certain circumstances, which is known to be problematic. See https://v8.dev/blog/math-random for details.
npm WARN deprecated acorn-dynamic-import@3.0.0: This is probably built in to whatever tool you're using. If you still need it... idk
npm WARN deprecated consolidate@0.15.1: Please upgrade to consolidate v1.0.0+ as it has been modernized with several long-awaited fixes implemented. Maintenance is supported by Forward Email at https://forwardemail.net/ ; follow/watch https://github.com/ladjs/consolidate for updates and release changelog
npm WARN deprecated urix@0.1.0: Please see https://github.com/lydell/urix#deprecated
npm WARN deprecated source-map-url@0.4.1: See https://github.com/lydell/source-map-url#deprecated
npm WARN deprecated resolve-url@0.2.1: https://github.com/lydell/resolve-url#deprecated
npm WARN deprecated source-map-resolve@0.5.3: See https://github.com/lydell/source-map-resolve#deprecated
npm WARN deprecated @babel/preset-stage-3@7.0.0-beta.54: 👋 We've deprecated any official stage presets in favor of users explicitly opt-ing into the proposals they want to use versus any perceived convenience. You can also check https://github.com/babel/babel/issues/7770 for more information.
npm WARN deprecated joi@11.4.0: This version has been deprecated in accordance with the hapi support policy (hapi.im/support). Please upgrade to the latest version to get the best features, bug fixes, and security patches. If you are unable to upgrade at this time, paid support is available for older versions (hapi.im/commercial).
npm WARN deprecated @babel/preset-stage-2@7.0.0-beta.54: 👋 We've deprecated any official stage presets in favor of users explicitly opt-ing into the proposals they want to use versus any perceived convenience. You can also check https://github.com/babel/babel/issues/7770 for more information.
npm WARN deprecated html-webpack-plugin@3.2.0: 3.x is no longer supported
npm WARN deprecated chokidar@2.0.4: Chokidar 2 does not receive security updates since 2019. Upgrade to chokidar 3 with 15x fewer dependencies
npm WARN deprecated uglify-es@3.3.9: support for ECMAScript is superseded by uglify-js as of v3.13.0
npm WARN deprecated svgo@1.3.2: This SVGO version is no longer supported. Upgrade to v2.x.x.
npm WARN deprecated core-js@2.6.12: core-js@<3.23.3 is no longer maintained and not recommended for usage due to the number of issues. Because of the V8 engine whims, feature detection in old core-js versions could cause a slowdown up to 100x even if nothing is polyfilled. Some versions have web compatibility issues. Please, upgrade your dependencies to the actual version of core-js.

added 1142 packages in 53s
intermediate containerRemoving intermediate container db55646cd7b6
---> 847f4ff28bf0
Step 15/23 : COPY ./src/router/* /app/src/router/
---> ea6bf96078c2
Step 16/23 : RUN ls /app/src/router/ -l
---> Running in a9fe332926fc
total 8
-rw-rw-r-- 1 root root 1029 Jul 26 15:31 index.ts
-rw-rw-r-- 1 root root 1040 Jul 26 15:31 routes.ts
Removing intermediate container a9fe332926fc
---> 36732835d125
Step 17/23 : RUN chmod 777 /app/src/router/index.ts
---> Running in 3455538a5b1d
Removing intermediate container 3455538a5b1d
---> 0d268325027a
Step 18/23 : RUN quasar build
---> Running in a9fbe6d0a39c

Build mode........ spa
Quasar theme...... mat
Quasar CLI........ v0.17.26
Quasar Framework.. v0.17.20
Debugging......... no

app:ensure-dev-deps Pinning strip-ansi dependency... +0ms
app:spawn [sync] Running "npm install --save-dev strip-ansi@=3.0.1" +1ms

added 14 packages, and changed 2 packages in 2m
app:quasar-conf Reading quasar.conf.js +2m
node:internal/fs/read/utf8:20
handleErrorFromBinding({ errno, syscall, path });
^

[ Error: ENOENT: no such file or directory, open '/app/src/router/index.js'

utf8:20 readFileSyncUtf8
node:internal/fs/read/utf8:20:3

node:fs:467 Object.readFileSync
node:fs:467:12

legacy-validations.js:104 legacyValidations
[lib]/[quasar-cli]/lib/legacy-validations.js:104:16

quasar-config.js:655 QuasarConfig.compile
[lib]/[quasar-cli]/lib/quasar-config.js:655:5

quasar-build:120 build
[lib]/[quasar-cli]/bin/quasar-build:120:16

task_queues:95 process.processTicksAndRejections
node:internal/process/task_queues:95:5

] {
errno: -2,
syscall: 'open',
code: 'ENOENT',
path: '/app/src/router/index.js'
}

Node.js v20.5.0

I have check /app/src/router/index.js exist in the path by RUN ls /app/src/router -al and RUN chmod 777 /app/src/router/index.ts, why quasar can't open it? any idea?

the Dockfile is

develop stage
FROM node:latest as develop-stage
WORKDIR /app
COPY electron-v7.3.3-linux-arm64.zip /app3/node_modules/electron/electron/releases/download/v7.3.3/
COPY SHASUMS256.txt /app3/node_modules/electron/electron/releases/download/v7.3.3/
COPY package*.json ./
#RUN npm i -g @quasar/cli
COPY . .
RUN ls /app/src/router -al
#COPY ./. /app/
#COPY ./. ./

build stage
FROM develop-stage as build-stage
RUN npm config set registry http://registry.npmmirror.com/
#RUN npm --registry=https://registry.npm.taobao.org/ install -g npm@latest
RUN export ELECTRON_MIRROR=$ELECTRON_MIRROR:http://npm.taobao.org/mirrors/electron/
RUN export ELECTRON_CONFIG_CACHE=$ELECTRON_CONFIG_CACHE:/app3/node_modules/electron/electron/releases/download/v7.3.3/
RUN npm --registry=http://registry.npmmirror.com/ install
RUN npm config get registry

#RUN curl -v https://registry.npmjs.com/
#RUN npm config set registry http://registry.npmjs.org/
#RUN npm cache clean --force
RUN npm install -g quasar-cli
COPY ./src/router/* /app/src/router/
RUN ls /app/src/router/ -l
RUN chmod 777 /app/src/router/index.ts
RUN quasar build

production stage
FROM nginx:latest as production-stage
RUN apt-get install -y tzdata
COPY --from=build-stage /app/dist/spa /usr/share/nginx/html
EXPOSE 80
CMD ["nginx", "-g", "daemon off;"]

**Comment by pdanpdan** [comment]
.js or .ts?

**Comment by gjtjx** [comment]
oh~~,but where is the index.js, can it generate from ts?
